README.md
# reggie

Run and encode simple register machine programs. Inspired by the COMP50003 Models of Computation course at Imperial College London.

## Install

To build `reggie` you need `alex` and `happy`. These can be installed by running: 

```
stack install alex happy
```

Installation of `reggie` is done using stack. Clone this repository and run:

```
cd reggie
make install
```

## Usage

To run a program:

```
reggie run my_program
reggie run my_program2 10 12
```

To encode a program:

```
reggie encode my_program
```

To decode a program:

```
reggie decode 1356938545749799165120738728340994364585019113312996743434322372066314551816
```

## Example programs

`examples/prog0` registers 0 and 1 to values of 3 and 4 respectively:

```
L0 : R0+ -> L1
L1 : R0+ -> L2
L2 : R0+ -> L4

L4 : R1+ -> L5
L5 : R1+ -> L6
L6 : R1+ -> L7
L7 : R1+ -> L8
L8 : HALT
```

This program encodes to 1356938545749799165120738728340994364585019113312996743434322372066314551816

`examples/prog1` adds 3 and 4 and stores the result in R0:

```
L0: R1+ -> L1
L1: R1+ -> L2
L2: R1+ -> L3

L3: R2+ -> L4
L4: R2+ -> L5
L5: R2+ -> L6
L6: R2+ -> L7

L7: R1- -> L8, L9
L8: R0+ -> L7
L9: R0+ -> L10

L10: R2- -> L11, L12
L11: R0+ -> L10
L12: R0+ -> L13

L13: HALT
```

This program encodes to `examples/prog1.encode`, which is 1010111 digits long!

`examples/add` takes two values from R1 and R2 and adds them to R0

```
# clear return register
L0: R0- -> L0, L1

# increment the sumands
L1: R1+ -> L2
L2: R2+ -> L3

# add R1 to R0
L3: R1- -> L4, L5
L4: R0+ -> L3

# add R2 to R0
L5: R2- -> L6, L7
L6: R0+ -> L5

# increment the output
L7: R0+ -> L8

L8: HALT
```

```
$ reggie run add 0 75 25
[(0,100),(1,0),(2,0)]
```

# To Do

- [x] Erroneous halting
- [x] Decoding programs
- [x] Add support for custom initial register states
- [ ] More detailed parsing / lexical errors
- [ ] Functions / modules (Gadgets)
- [ ] Custom labels
- [x] Comments


ChangeLog.md
# Changelog for reggie

## Unreleased changes


test/EncodeSpec.hs
module EncodeSpec (spec) where

import Test.Hspec
import Encode
import Program (Program(..), Instruction(..), Label(..))
import Data.Map (fromList)

incr :: Integer -> Integer -> Instruction
incr reg next
  = Incr reg (Label next)

decr :: Integer -> (Integer, Integer) -> Instruction
decr reg (nextThen, nextElse)
  = Decr reg (Label nextThen, Label nextElse)

program :: [Instruction] -> Program
program
  = Program . fromList . zip [Label 0..]

spec :: Spec
spec
  = describe "Encode" $ do
      describe "encode" $ do
        it "can encode pairs" $ do
          encode (0 :: Int, 0 :: Int) `shouldBe` 1
          encode (1 :: Int, 0 :: Int) `shouldBe` 2
          encode (5 :: Int, 0 :: Int) `shouldBe` 32
          encode (0 :: Int, 3 :: Int) `shouldBe` 7
          encode (1 :: Int, 2 :: Int) `shouldBe` 10
          encode (2 :: Int, 1 :: Int) `shouldBe` 12
          encode (2 :: Int, 2 :: Int) `shouldBe` 20

        it "can encode instructions" $ do
          encode (incr 0 0)      `shouldBe` 1
          encode (incr 4 6)      `shouldBe` 3328
          encode (decr 0 (0, 0)) `shouldBe` 2
          encode (decr 3 (1, 8)) `shouldBe` 8576
          encode Halt            `shouldBe` 0

        it "can encode programs" $ do
          encode (program [Halt]) `shouldBe` 1
          encode (program [decr 0 (0, 2), Halt]) `shouldBe` 786432

      describe "encode'" $ do
         it "can encode pairs correctly" $ do
           encode' (0 :: Int, 0 :: Int) `shouldBe` 0
           encode' (1 :: Int, 0 :: Int) `shouldBe` 1
           encode' (5 :: Int, 0 :: Int) `shouldBe` 31
           encode' (0 :: Int, 3 :: Int) `shouldBe` 6
           encode' (1 :: Int, 2 :: Int) `shouldBe` 9
           encode' (2 :: Int, 1 :: Int) `shouldBe` 11
           encode' (2 :: Int, 2 :: Int) `shouldBe` 19

      describe "decode" $ do
        it "can decode pairs" $ do
          decode 1  `shouldBe` (0 :: Int, 0 :: Int)
          decode 2  `shouldBe` (1 :: Int, 0 :: Int)
          decode 32 `shouldBe` (5 :: Int, 0 :: Int)
          decode 7  `shouldBe` (0 :: Int, 3 :: Int)
          decode 10 `shouldBe` (1 :: Int, 2 :: Int)
          decode 12 `shouldBe` (2 :: Int, 1 :: Int)
          decode 20 `shouldBe` (2 :: Int, 2 :: Int)

        it "can decode instructions" $ do
          decode 1    `shouldBe` incr 0 0
          decode 3328 `shouldBe` incr 4 6
          decode 2    `shouldBe` decr 0 (0, 0)
          decode 8576 `shouldBe` decr 3 (1, 8)
          decode 0    `shouldBe` Halt

        it "can decode programs" $ do
          decode 1      `shouldBe` program [Halt]
          decode 786432 `shouldBe` program [decr 0 (0, 2), Halt]



test/ParserSpec.hs
module ParserSpec (spec) where

import Test.Hspec
import Parser
import Parser.Tokens
import Program (Instruction(..), Label(..))
import Data.Map (fromList)

spec :: Spec
spec
  = describe "Parser" $ do 
      describe "tokenize" $ do
        it "can remove comments" $
          tokenize "# this is a comment" `shouldBe` []

        it "can remove whitespace but keeps newlines" $
          tokenize "   \n   " `shouldBe` [TokenNewLine]

        it "can tokenize halt instructions" $
          tokenize "HALT" `shouldBe` [TokenHalt]

        it "can tokenize increment instructions" $
          tokenize "L0: R1+ -> L2" `shouldBe`
            [TokenLabel 0, TokenColon, TokenRegister 1, TokenIncr, TokenArrow, TokenLabel 2]

        it "can tokenize decremenet instructions" $
          tokenize "L1: R10- -> L43, L2" `shouldBe`
            [TokenLabel 1, TokenColon, TokenRegister 10, TokenDecr, TokenArrow, TokenLabel 43, TokenComma, TokenLabel 2]

      describe "parse" $ do
        it "can parse halt instructions" $
          parse [TokenLabel 4, TokenColon, TokenHalt] `shouldBe` fromList [(Label 4, Halt)]

        it "can parse increment instructions" $
          parse [TokenLabel 0, TokenColon, TokenRegister 1, TokenIncr, TokenArrow, TokenLabel 2]
            `shouldBe` fromList [(Label 0, Incr 1 (Label 2))]

        it "can parse decrement instructions" $
          parse [TokenLabel 1, TokenColon, TokenRegister 10, TokenDecr, TokenArrow, TokenLabel 43, TokenComma, TokenLabel 2]
            `shouldBe` fromList [(Label 1, Decr 10 (Label 43, Label 2))]


test/ProgramSpec.hs
module ProgramSpec (spec) where

import Test.Hspec
import Program (Program(..), Instruction(..), Label(..), run)
import Data.Map (fromList)

incr :: Integer -> Integer -> Instruction
incr reg next
  = Incr reg (Label next)

decr :: Integer -> (Integer, Integer) -> Instruction
decr reg (nextThen, nextElse)
  = Decr reg (Label nextThen, Label nextElse)

program :: [Instruction] -> Program
program
  = Program . fromList . zip [Label 0..]

spec :: Spec
spec
  = do
    describe "run" $ do
      it "can halt" $
        run (program [Halt]) `shouldBe` []

      it "can increment values" $ do
        run (program [incr 0 1, Halt]) `shouldBe` [(0, 1)]
        run (program [incr 1 1, Halt]) `shouldBe` [(1, 1)]
        run (program [incr 0 1, incr 0 2, Halt]) `shouldBe` [(0, 2)]
        run (program [incr 0 1, incr 1 2, Halt]) `shouldBe` [(0, 1), (1, 1)]

      it "can decrement values" $ do
        run (program [decr 0 (1,1), Halt]) `shouldBe` [(0, -1)]
        run (program [decr 0 (1,1), decr 0 (2, 2), Halt]) `shouldBe` [(0, -2)]
        run (program [decr 0 (1,1), decr 1 (2, 2), Halt]) `shouldBe` [(0, -1), (1, -1)]

      it "can branch on decrement" $ do
        run (program [decr 0 (1, 2), incr 1 2, Halt]) `shouldBe` [(0, -1)]
        run (program [incr 0 1, decr 0 (2, 3), incr 1 0, Halt]) `shouldBe` [(0, 0)]
        run (program [incr 0 1, incr 0 2, incr 0 3, decr 0 (4, 5), incr 1 3, Halt])
          `shouldBe` [(0, 0), (1, 2)]


src/Encode.hs
{-# LANGUAGE FlexibleInstances #-}

module Encode where

import Program (Instruction(..), Program(..), Label(..))
import Data.List (unfoldr)
import Data.Map (toAscList, fromList)

class Encode a where
 -- Bijection from NxN to N+
  encode :: Integral b => a -> b

  decode :: Integral b => b -> a

  -- Bijection from NxN to N
  encode' :: Integral b => a -> b
  encode' = pred . encode

  decode' :: Integral b => b -> a
  decode' = decode . succ

logBaseIntegral :: Integral a => a -> a -> a
logBaseIntegral base x
  | x <= 1    = 0
  | otherwise = 1 + logBaseIntegral base (x `div` base)

instance (Integral a, Integral b) => Encode (a, b) where
  encode (x, y)
    = fromIntegral $ (2 ^ x) * (2 * y + 1)

  decode z
    = let q = until odd (`div` 2) z
          y = fromIntegral $ (q - 1) `div` 2
          x = fromIntegral $ logBaseIntegral 2 (z `div` q)
       in (x, y)

instance Encode a => Encode [a] where
  encode
    = foldr (\x y -> encode (encode x, y)) 0

  decode z
    = let f 0 = Nothing
          f a = let (x, y) = decode a
                 in Just (decode x, y)
       in unfoldr f z

instance Encode Instruction where
  encode (Incr reg (Label next))
    = encode (2 * reg, next)
  encode (Decr reg (Label nextThen, Label nextElse))
    = encode (2 * reg + 1, encode' (nextThen, nextElse))
  encode Halt
    = 0

  decode 0
    = Halt
  decode x
     = let (y,z) = decode x
           i = y `div` 2
           (j,k) = decode' z
        in if even y
              then Incr i (Label z)
              else Decr i (Label j, Label k)

instance Encode Program where
  encode (Program instrs)
    = encode $ padMissing (Label 0) $ toAscList instrs
      where
        padMissing :: Label -> [(Label, Instruction)] -> [Instruction]
        padMissing _ []
          = []
        padMissing label pairs@((label', instr) : remaining)
          | label < label' = Halt : padMissing (succ label) pairs
          | otherwise      = instr : padMissing (succ label') remaining

  decode x
    = Program $ fromList $ zipWith (\l i -> (Label l, i)) [0..] $ decode x


src/Program.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

module Program where

import Prelude hiding (lookup)
import Data.Map (Map(..), alter, fromList, lookup, toAscList)
import Data.Maybe (fromMaybe)

newtype Label
  = Label Integer
    deriving (Eq, Ord, Enum)

instance Show Label where
  show (Label l)
    = "L" ++ show l

data Instruction
  = Incr Integer Label
  | Decr Integer (Label, Label)
  | Halt
    deriving Eq

instance Show Instruction where
  show (Incr reg next)
    = "R" ++ show reg ++ "+ -> " ++ show next
  show (Decr reg (nextThen, nextElse))
    = "R" ++ show reg ++ "- -> " ++ show nextThen ++ ", " ++ show nextElse
  show Halt
    = "HALT"

newtype Program
  = Program (Map Label Instruction)
    deriving Eq

instance Show Program where
  show (Program program)
    = unlines $ map (\(l, i) -> show l ++ ": " ++ show i) $ toAscList program

data Machine
  = Machine Program (Map Integer Integer) Label

start :: Label
start = Label 0

run :: Program -> [(Integer, Integer)]
run program
  = runMem program []

runMem :: Program -> [(Integer, Integer)] -> [(Integer, Integer)]
runMem program memoryList
  = let memory = fromList memoryList
        (Machine _ memory' _) = run' $ Machine program memory start
     in toAscList memory'

run' :: Machine -> Machine
run' machine@(Machine program _ pc)
  = case getInstr program pc of
      Halt  -> machine
      instr -> run' $ step instr machine

step :: Instruction -> Machine -> Machine
step (Incr reg next) (Machine program memory _)
  = Machine program (alter increment reg memory) next
step (Decr reg nexts) (Machine program memory _)
  = let memory' = alter decrement reg memory
        next = pickNext (regPositive reg memory') nexts
     in Machine program memory' next
step' Halt machine
  = machine

getInstr :: Program -> Label -> Instruction
getInstr (Program program) label
  = fromMaybe Halt $ lookup label program

regPositive :: Integer -> Map Integer Integer -> Bool
regPositive reg memory
  = case lookup reg memory of
      Nothing -> False
      Just x  -> x > 0

pickNext :: Bool -> (Label, Label) -> Label
pickNext True (x, _)  = x
pickNext False (_, y) = y

increment :: Maybe Integer -> Maybe Integer
increment Nothing
  = Just 1
increment (Just x)
  = Just $ succ x

decrement :: Maybe Integer -> Maybe Integer
decrement Nothing
  = Just (-1)
decrement (Just x)
  = Just $ pred x


src/Parser/Grammer.hs
{-# OPTIONS_GHC -w #-}
module Parser.Grammer where
import Parser.Tokens
import Program
import qualified Data.Array as Happy_Data_Array
import qualified Data.Bits as Bits
import Control.Applicative(Applicative(..))
import Control.Monad (ap)

-- parser produced by Happy Version 1.19.12

data HappyAbsSyn t4 t5
	= HappyTerminal (Token)
	| HappyErrorToken Int
	| HappyAbsSyn4 t4
	| HappyAbsSyn5 t5

happyExpList :: Happy_Data_Array.Array Int Int
happyExpList = Happy_Data_Array.listArray (0,46) ([8800,304,2048,384,8,0,0,24640,34,38912,17416,4096,2048,256,128,3072,0,0,32,16,0,32832,16384,8192,0,0,2,512,0,0,0
	])

{-# NOINLINE happyExpListPerState #-}
happyExpListPerState st =
    token_strs_expected
  where token_strs = ["error","%dummy","%start_parse","Program","Instruction","register","label","colon","arrow","halt","incr","decr","comma","'\\n'","%eof"]
        bit_start = st * 15
        bit_end = (st + 1) * 15
        read_bit = readArrayBit happyExpList
        bits = map read_bit [bit_start..bit_end - 1]
        bits_indexed = zip bits [0..14]
        token_strs_expected = concatMap f bits_indexed
        f (False, _) = []
        f (True, nr) = [token_strs !! nr]

action_0 (6) = happyShift action_3
action_0 (7) = happyShift action_4
action_0 (10) = happyShift action_5
action_0 (14) = happyShift action_8
action_0 (4) = happyGoto action_6
action_0 (5) = happyGoto action_7
action_0 _ = happyReduce_4

action_1 (6) = happyShift action_3
action_1 (7) = happyShift action_4
action_1 (10) = happyShift action_5
action_1 (5) = happyGoto action_2
action_1 _ = happyFail (happyExpListPerState 1)

action_2 (14) = happyShift action_10
action_2 _ = happyFail (happyExpListPerState 2)

action_3 (11) = happyShift action_12
action_3 (12) = happyShift action_13
action_3 _ = happyFail (happyExpListPerState 3)

action_4 (8) = happyShift action_11
action_4 _ = happyFail (happyExpListPerState 4)

action_5 _ = happyReduce_10

action_6 (15) = happyAccept
action_6 _ = happyFail (happyExpListPerState 6)

action_7 (14) = happyShift action_10
action_7 _ = happyReduce_3

action_8 (6) = happyShift action_3
action_8 (7) = happyShift action_4
action_8 (10) = happyShift action_5
action_8 (14) = happyShift action_8
action_8 (4) = happyGoto action_9
action_8 (5) = happyGoto action_7
action_8 _ = happyReduce_4

action_9 _ = happyReduce_2

action_10 (6) = happyShift action_3
action_10 (7) = happyShift action_4
action_10 (10) = happyShift action_5
action_10 (14) = happyShift action_8
action_10 (4) = happyGoto action_18
action_10 (5) = happyGoto action_7
action_10 _ = happyReduce_4

action_11 (6) = happyShift action_16
action_11 (10) = happyShift action_17
action_11 _ = happyFail (happyExpListPerState 11)

action_12 (9) = happyShift action_15
action_12 _ = happyFail (happyExpListPerState 12)

action_13 (9) = happyShift action_14
action_13 _ = happyFail (happyExpListPerState 13)

action_14 (7) = happyShift action_22
action_14 _ = happyFail (happyExpListPerState 14)

action_15 (7) = happyShift action_21
action_15 _ = happyFail (happyExpListPerState 15)

action_16 (11) = happyShift action_19
action_16 (12) = happyShift action_20
action_16 _ = happyFail (happyExpListPerState 16)

action_17 _ = happyReduce_7

action_18 _ = happyReduce_1

action_19 (9) = happyShift action_25
action_19 _ = happyFail (happyExpListPerState 19)

action_20 (9) = happyShift action_24
action_20 _ = happyFail (happyExpListPerState 20)

action_21 _ = happyReduce_8

action_22 (13) = happyShift action_23
action_22 _ = happyFail (happyExpListPerState 22)

action_23 (7) = happyShift action_28
action_23 _ = happyFail (happyExpListPerState 23)

action_24 (7) = happyShift action_27
action_24 _ = happyFail (happyExpListPerState 24)

action_25 (7) = happyShift action_26
action_25 _ = happyFail (happyExpListPerState 25)

action_26 _ = happyReduce_5

action_27 (13) = happyShift action_29
action_27 _ = happyFail (happyExpListPerState 27)

action_28 _ = happyReduce_9

action_29 (7) = happyShift action_30
action_29 _ = happyFail (happyExpListPerState 29)

action_30 _ = happyReduce_6

happyReduce_1 = happySpecReduce_3  4 happyReduction_1
happyReduction_1 (HappyAbsSyn4  happy_var_3)
	_
	(HappyAbsSyn5  happy_var_1)
	 =  HappyAbsSyn4
		 (happy_var_1 : happy_var_3
	)
happyReduction_1 _ _ _  = notHappyAtAll 

happyReduce_2 = happySpecReduce_2  4 happyReduction_2
happyReduction_2 (HappyAbsSyn4  happy_var_2)
	_
	 =  HappyAbsSyn4
		 (happy_var_2
	)
happyReduction_2 _ _  = notHappyAtAll 

happyReduce_3 = happySpecReduce_1  4 happyReduction_3
happyReduction_3 (HappyAbsSyn5  happy_var_1)
	 =  HappyAbsSyn4
		 ([happy_var_1]
	)
happyReduction_3 _  = notHappyAtAll 

happyReduce_4 = happySpecReduce_0  4 happyReduction_4
happyReduction_4  =  HappyAbsSyn4
		 ([]
	)

happyReduce_5 = happyReduce 6 5 happyReduction_5
happyReduction_5 ((HappyTerminal (TokenLabel happy_var_6)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenRegister happy_var_3)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	happyRest)
	 = HappyAbsSyn5
		 (Incr happy_var_3 happy_var_6
	) `HappyStk` happyRest

happyReduce_6 = happyReduce 8 5 happyReduction_6
happyReduction_6 ((HappyTerminal (TokenLabel happy_var_8)) `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenLabel happy_var_6)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenRegister happy_var_3)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	happyRest)
	 = HappyAbsSyn5
		 (Decr happy_var_3 (happy_var_6, happy_var_8)
	) `HappyStk` happyRest

happyReduce_7 = happySpecReduce_3  5 happyReduction_7
happyReduction_7 _
	_
	_
	 =  HappyAbsSyn5
		 (Halt
	)

happyReduce_8 = happyReduce 4 5 happyReduction_8
happyReduction_8 ((HappyTerminal (TokenLabel happy_var_4)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenRegister happy_var_1)) `HappyStk`
	happyRest)
	 = HappyAbsSyn5
		 (Incr happy_var_1 happy_var_4
	) `HappyStk` happyRest

happyReduce_9 = happyReduce 6 5 happyReduction_9
happyReduction_9 ((HappyTerminal (TokenLabel happy_var_6)) `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenLabel happy_var_4)) `HappyStk`
	_ `HappyStk`
	_ `HappyStk`
	(HappyTerminal (TokenRegister happy_var_1)) `HappyStk`
	happyRest)
	 = HappyAbsSyn5
		 (Decr happy_var_1 (happy_var_4, happy_var_6)
	) `HappyStk` happyRest

happyReduce_10 = happySpecReduce_1  5 happyReduction_10
happyReduction_10 _
	 =  HappyAbsSyn5
		 (Halt
	)

happyNewToken action sts stk [] =
	action 15 15 notHappyAtAll (HappyState action) sts stk []

happyNewToken action sts stk (tk:tks) =
	let cont i = action i i tk (HappyState action) sts stk tks in
	case tk of {
	TokenRegister happy_dollar_dollar -> cont 6;
	TokenLabel happy_dollar_dollar -> cont 7;
	TokenColon -> cont 8;
	TokenArrow -> cont 9;
	TokenHalt -> cont 10;
	TokenIncr -> cont 11;
	TokenDecr -> cont 12;
	TokenComma -> cont 13;
	TokenNewLine -> cont 14;
	_ -> happyError' ((tk:tks), [])
	}

happyError_ explist 15 tk tks = happyError' (tks, explist)
happyError_ explist _ tk tks = happyError' ((tk:tks), explist)

newtype HappyIdentity a = HappyIdentity a
happyIdentity = HappyIdentity
happyRunIdentity (HappyIdentity a) = a

instance Functor HappyIdentity where
    fmap f (HappyIdentity a) = HappyIdentity (f a)

instance Applicative HappyIdentity where
    pure  = HappyIdentity
    (<*>) = ap
instance Monad HappyIdentity where
    return = pure
    (HappyIdentity p) >>= q = q p

happyThen :: () => HappyIdentity a -> (a -> HappyIdentity b) -> HappyIdentity b
happyThen = (>>=)
happyReturn :: () => a -> HappyIdentity a
happyReturn = (return)
happyThen1 m k tks = (>>=) m (\a -> k a tks)
happyReturn1 :: () => a -> b -> HappyIdentity a
happyReturn1 = \a tks -> (return) a
happyError' :: () => ([(Token)], [String]) -> HappyIdentity a
happyError' = HappyIdentity . (\(tokens, _) -> parseError tokens)
parse tks = happyRunIdentity happySomeParser where
 happySomeParser = happyThen (happyParse action_0 tks) (\x -> case x of {HappyAbsSyn4 z -> happyReturn z; _other -> notHappyAtAll })

happySeq = happyDontSeq


parseError :: [Token] -> a
parseError _ = error "Parse error"
{-# LINE 1 "templates/GenericTemplate.hs" #-}
-- $Id: GenericTemplate.hs,v 1.26 2005/01/14 14:47:22 simonmar Exp $










































data Happy_IntList = HappyCons Int Happy_IntList








































infixr 9 `HappyStk`
data HappyStk a = HappyStk a (HappyStk a)

-----------------------------------------------------------------------------
-- starting the parse

happyParse start_state = happyNewToken start_state notHappyAtAll notHappyAtAll

-----------------------------------------------------------------------------
-- Accepting the parse

-- If the current token is ERROR_TOK, it means we've just accepted a partial
-- parse (a %partial parser).  We must ignore the saved token on the top of
-- the stack in this case.
happyAccept (1) tk st sts (_ `HappyStk` ans `HappyStk` _) =
        happyReturn1 ans
happyAccept j tk st sts (HappyStk ans _) = 
         (happyReturn1 ans)

-----------------------------------------------------------------------------
-- Arrays only: do the next action









































indexShortOffAddr arr off = arr Happy_Data_Array.! off


{-# INLINE happyLt #-}
happyLt x y = (x < y)






readArrayBit arr bit =
    Bits.testBit (indexShortOffAddr arr (bit `div` 16)) (bit `mod` 16)






-----------------------------------------------------------------------------
-- HappyState data type (not arrays)



newtype HappyState b c = HappyState
        (Int ->                    -- token number
         Int ->                    -- token number (yes, again)
         b ->                           -- token semantic value
         HappyState b c ->              -- current state
         [HappyState b c] ->            -- state stack
         c)



-----------------------------------------------------------------------------
-- Shifting a token

happyShift new_state (1) tk st sts stk@(x `HappyStk` _) =
     let i = (case x of { HappyErrorToken (i) -> i }) in
--     trace "shifting the error token" $
     new_state i i tk (HappyState (new_state)) ((st):(sts)) (stk)

happyShift new_state i tk st sts stk =
     happyNewToken new_state ((st):(sts)) ((HappyTerminal (tk))`HappyStk`stk)

-- happyReduce is specialised for the common cases.

happySpecReduce_0 i fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happySpecReduce_0 nt fn j tk st@((HappyState (action))) sts stk
     = action nt j tk st ((st):(sts)) (fn `HappyStk` stk)

happySpecReduce_1 i fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happySpecReduce_1 nt fn j tk _ sts@(((st@(HappyState (action))):(_))) (v1`HappyStk`stk')
     = let r = fn v1 in
       happySeq r (action nt j tk st sts (r `HappyStk` stk'))

happySpecReduce_2 i fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happySpecReduce_2 nt fn j tk _ ((_):(sts@(((st@(HappyState (action))):(_))))) (v1`HappyStk`v2`HappyStk`stk')
     = let r = fn v1 v2 in
       happySeq r (action nt j tk st sts (r `HappyStk` stk'))

happySpecReduce_3 i fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happySpecReduce_3 nt fn j tk _ ((_):(((_):(sts@(((st@(HappyState (action))):(_))))))) (v1`HappyStk`v2`HappyStk`v3`HappyStk`stk')
     = let r = fn v1 v2 v3 in
       happySeq r (action nt j tk st sts (r `HappyStk` stk'))

happyReduce k i fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happyReduce k nt fn j tk st sts stk
     = case happyDrop (k - ((1) :: Int)) sts of
         sts1@(((st1@(HappyState (action))):(_))) ->
                let r = fn stk in  -- it doesn't hurt to always seq here...
                happyDoSeq r (action nt j tk st1 sts1 r)

happyMonadReduce k nt fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happyMonadReduce k nt fn j tk st sts stk =
      case happyDrop k ((st):(sts)) of
        sts1@(((st1@(HappyState (action))):(_))) ->
          let drop_stk = happyDropStk k stk in
          happyThen1 (fn stk tk) (\r -> action nt j tk st1 sts1 (r `HappyStk` drop_stk))

happyMonad2Reduce k nt fn (1) tk st sts stk
     = happyFail [] (1) tk st sts stk
happyMonad2Reduce k nt fn j tk st sts stk =
      case happyDrop k ((st):(sts)) of
        sts1@(((st1@(HappyState (action))):(_))) ->
         let drop_stk = happyDropStk k stk





             _ = nt :: Int
             new_state = action

          in
          happyThen1 (fn stk tk) (\r -> happyNewToken new_state sts1 (r `HappyStk` drop_stk))

happyDrop (0) l = l
happyDrop n ((_):(t)) = happyDrop (n - ((1) :: Int)) t

happyDropStk (0) l = l
happyDropStk n (x `HappyStk` xs) = happyDropStk (n - ((1)::Int)) xs

-----------------------------------------------------------------------------
-- Moving to a new state after a reduction









happyGoto action j tk st = action j j tk (HappyState action)


-----------------------------------------------------------------------------
-- Error recovery (ERROR_TOK is the error token)

-- parse error if we are in recovery and we fail again
happyFail explist (1) tk old_st _ stk@(x `HappyStk` _) =
     let i = (case x of { HappyErrorToken (i) -> i }) in
--      trace "failing" $ 
        happyError_ explist i tk

{-  We don't need state discarding for our restricted implementation of
    "error".  In fact, it can cause some bogus parses, so I've disabled it
    for now --SDM

-- discard a state
happyFail  ERROR_TOK tk old_st CONS(HAPPYSTATE(action),sts) 
                                                (saved_tok `HappyStk` _ `HappyStk` stk) =
--      trace ("discarding state, depth " ++ show (length stk))  $
        DO_ACTION(action,ERROR_TOK,tk,sts,(saved_tok`HappyStk`stk))
-}

-- Enter error recovery: generate an error token,
--                       save the old token and carry on.
happyFail explist i tk (HappyState (action)) sts stk =
--      trace "entering error recovery" $
        action (1) (1) tk (HappyState (action)) sts ((HappyErrorToken (i)) `HappyStk` stk)

-- Internal happy errors:

notHappyAtAll :: a
notHappyAtAll = error "Internal Happy error\n"

-----------------------------------------------------------------------------
-- Hack to get the typechecker to accept our action functions







-----------------------------------------------------------------------------
-- Seq-ing.  If the --strict flag is given, then Happy emits 
--      happySeq = happyDoSeq
-- otherwise it emits
--      happySeq = happyDontSeq

happyDoSeq, happyDontSeq :: a -> b -> b
happyDoSeq   a b = a `seq` b
happyDontSeq a b = b

-----------------------------------------------------------------------------
-- Don't inline any functions from the template.  GHC has a nasty habit
-- of deciding to inline happyGoto everywhere, which increases the size of
-- the generated parser quite a bit.









{-# NOINLINE happyShift #-}
{-# NOINLINE happySpecReduce_0 #-}
{-# NOINLINE happySpecReduce_1 #-}
{-# NOINLINE happySpecReduce_2 #-}
{-# NOINLINE happySpecReduce_3 #-}
{-# NOINLINE happyReduce #-}
{-# NOINLINE happyMonadReduce #-}
{-# NOINLINE happyGoto #-}
{-# NOINLINE happyFail #-}

-- end of Happy Template.
